bundle/filefs.go

```diff
@@ -0,0 +1,73 @@
+// +build go1.16
+
+package bundle
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type dirLoaderFS struct {
+	filesystem fs.FS
+	files      []string
+	idx        int
+}
+
+// Interface guard for dirLoaderFS to validate that it implements the DirectoryLoader interface
+var _ DirectoryLoader = (*dirLoader)(nil)
+
+// NewFSLoader returns a basic DirectoryLoader implementation
+// that will load files from a fs.FS interface
+func NewFSLoader(filesystem fs.FS) DirectoryLoader {
+	d := dirLoaderFS{
+		filesystem: filesystem,
+	}
+
+	return &d
+}
+
+// NextFile iterates to the next file in the directory tree
+// and returns a file Descriptor for the file.
+func (d *dirLoaderFS) NextFile() (*Descriptor, error) {
+	// build a list of all files we will iterate over and read, but only one time
+	if d.files == nil {
```

Comment:
We'll never call this multiple times concurrently, I guess? If invoked twice, this would be problematic: either one thread could read and write this...

Reply:
Yep! But this is an issue with the current NewDirectoryLoader as well. Verified it with a test:

```Go
func TestDirectoryLoaderTwice(t *testing.T) {
	test.WithTempFS(archiveFiles, func(rootDir string) {
		loader := NewDirectoryLoader(rootDir)

		testLoader(t, loader, rootDir, archiveFiles)
		testLoader(t, loader, rootDir, archiveFiles)
	})
}
```

```shell
go test -timeout 30s -run ^TestDirectoryLoaderTwice$ github.com/open-policy-agent/opa/bundle                                                                                                                             22:55:49
--- FAIL: TestDirectoryLoaderTwice (0.00s)
    file_test.go:78: Expected to read 6 files, read 0
FAIL
FAIL    github.com/open-policy-agent/opa/bundle 0.004s
FAIL
```

It will always hit the following if the files have been read:
```Go
	// If done reading files then just return io.EOF
	// errors for each NextFile() call
	if d.idx >= len(d.files) {
		return nil, io.EOF
	}
```

Would you like to keep it as is or do something else?

Reply:
Yeah, same story there: https://github.com/open-policy-agent/opa/blob/27c8d75d3528dc8a04e716d5455f97ee7fb537bc/bundle/file.go#L112-L113

Btw what I meant was _concurrent access_, not reading the files twice. So, a test case to show what's going on would be this:

```
func TestDirectoryLoaderTwice(t *testing.T) {
	test.WithTempFS(archiveFiles, func(rootDir string) {
		loader := NewDirectoryLoader(rootDir)

		wg := new(sync.WaitGroup)
		wg.Add(2)
		go func() {
			testLoader(t, loader, rootDir, archiveFiles)
			wg.Done()
		}()
		go func() {
			testLoader(t, loader, rootDir, archiveFiles)
			wg.Done()
		}()
		wg.Wait()
	})
}
```

...when testing with the race detector on, (`go test ./bundle -run ^TestDirectoryLoaderTwice$ -v -race -count=1`), it will yield this:
```
WARNING: DATA RACE
Read at 0x00c0001d67f0 by goroutine 9:
  github.com/open-policy-agent/opa/bundle.(*dirLoader).NextFile()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file.go:112 +0x64
  github.com/open-policy-agent/opa/bundle.testLoader()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:79 +0x83
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice.func1.2()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:192 +0x84

Previous write at 0x00c0001d67f0 by goroutine 8:
  github.com/open-policy-agent/opa/bundle.(*dirLoader).NextFile()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file.go:113 +0x84e
  github.com/open-policy-agent/opa/bundle.testLoader()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:79 +0x83
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice.func1.1()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:188 +0x84

Goroutine 9 (running) created at:
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice.func1()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:191 +0x20f
  github.com/open-policy-agent/opa/util/test.WithTempFS()
      /Users/stephan/Sources/open-policy-agent/opa-main/util/test/tempfs.go:21 +0xeb
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:182 +0x72
  testing.tRunner()
      /nix/store/398g6wfmd06cvqzkbjhccwadfb5mwlk2-go-1.16.4/share/go/src/testing/testing.go:1193 +0x202

Goroutine 8 (running) created at:
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice.func1()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:187 +0x1b3
  github.com/open-policy-agent/opa/util/test.WithTempFS()
      /Users/stephan/Sources/open-policy-agent/opa-main/util/test/tempfs.go:21 +0xeb
  github.com/open-policy-agent/opa/bundle.TestDirectoryLoaderTwice()
      /Users/stephan/Sources/open-policy-agent/opa-main/bundle/file_test.go:182 +0x72
  testing.tRunner()
      /nix/store/398g6wfmd06cvqzkbjhccwadfb5mwlk2-go-1.16.4/share/go/src/testing/testing.go:1193 +0x202
```

So that confirms that it's not safe for concurrent use -- but since that's the existing implementation, too, we're surely not using it concurrently. Hence the same behaviour in this addition won't do harm.

(If you'd like to avoid it, adding a mutex would be a way to go.)

Reply:
I've added a mutex and lock/unlock. Do you think this is enough?